### PR TITLE
Do not pause rendering when android activity loses focus

### PR DESCRIPTION
**Pull request by jonahwilliams**
Originally the `FlutterActivity` used `Activity#onPause()` and `Activity#onResume()` as a proxy for whether the Android activity was visible.  This made perfect sense before splitscreen, picture in picture, and chromebooks.  Currently, any time a Flutter activity loses focus, rendering will stop.  This is more obvious when there is an animation playing.  I haven't had time to check on a chromebook, but I imagine that losing focus there will stop rendering too. 

`onPause()` will instead trigger the `inactive` state on Android, previously only used for iOS.  This corresponds to foreground visible but not focused, which matches the current behavior.  since the `paused` lifecycle is also used by iOS, this will still be used - but only `onStop()` for android.  This is called whenever an activity goes into the background or is completely invisible. 

#### Picture in Picture
Original on left, with fix on right.
gallery_pipgallery_working_pip

#### Splitscreen 
Original on left, with fix on right.
gallery_splitgallery_working_split

### Follow up work
Currently Flutter applications will still restart when entering splitscreen mode.  The fix for this is separate, just an addition to the `android:configChanges` entry.

Fixes https://github.com/flutter/flutter/issues/11906, https://github.com/flutter/flutter/issues/14619, https://github.com/flutter/flutter/issues/14665

### CODE REVIEW

**Comment by Hixie** [comment]
What does this do to the app lifecycle events?
https://master-docs-flutter-io.firebaseapp.com/flutter/services/SystemChannels/lifecycle-constant.html
https://master-docs-flutter-io.firebaseapp.com/flutter/dart-ui/AppLifecycleState-class.html

The theory is that the framework is set up to be entirely in charge of when frames should be pumped; the engine should schedule frames every 16ms so long as the framework is requesting frames.

**Comment by jonahwilliams** [comment]
We'll need to update them.  I can actually move the logic to the `onStop` and `onRestart` lifecycle handlers, then we can correctly stop asking for frames.

"Interestingly" if I don't add `screenLayout|density` to `android:configChanges`, the app crashes (with the above changes, not with PR changes).

Investigating

**Comment by jonahwilliams** [comment]
I also found the code in `scheduler/binding` - that should just be updated and the engine case for pausing frames adjusted.  We can move the existing suspending logic into `onStop`

**Comment by jonahwilliams** [comment]
I've found the cause of the assertion error (not a crash, caused by sending a message "platformChrome" when the native view doesn't exist).  It seems that restarting the App, caused by the change in layout/density, caused RunBundleAndSnapshot to be re-invoked

**Comment by jonahwilliams** [comment]
@Hixie I've updated the PR.  This change doesn't require anything from the flutter side, though we should consolidate the life cycles later.

**Comment by cbracken** [comment]
Generally, LGTM. /cc @jason-simmons for Android expertise.

Given that we're changing which events are fired and when, are there situations this will be a breaking change to users who rely on the current behaviour? i.e. is there any change in behaviour when not in splitscreen or using picture-in-picture mode?

File: shell/platform/android/io/flutter/view/FlutterNativeView.java [hunk 1]
```diff
@@ -128,7 +128,14 @@ private void attach(FlutterNativeView view) {
 
     // Called by native to send us a platform message.
     private void handlePlatformMessage(final String channel, byte[] message, final int replyId) {
-        assertAttached();
+        // The platform may not be attached immediately in certain cases where a new bundle is run -
+        // the native view is created in a separate thread. This mostly happens when the app restarts in dev
+        // mode when switching into split-screen mode. Preventing app restarts on layout and density
+        // changes will prevent this, and afterwards this can be changed back to an assert.
+        if (!isAttached()) {
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/view/FlutterNativeView.java]: Does this still happen if you apply the pending engine refactoring by @chinmaygarde ? (https://github.com/chinmaygarde/flutter_engine/tree/shell)

The Dart application may fail if it sends a message to the host and the host drops the message with no response.

Comment (jonahwilliams) [reply shell/platform/android/io/flutter/view/FlutterNativeView.java]: So looking into this more - I can reproduce the issue on master, somewhat flaky though.  Might only be a debug issue

File: shell/platform/android/io/flutter/view/FlutterView.java [hunk 2]
```diff
@@ -267,7 +267,7 @@ public void addActivityLifecycleListener(ActivityLifecycleListener listener) {
     }
 
     public void onPause() {
-        mFlutterLifecycleChannel.send("AppLifecycleState.paused");
+        mFlutterLifecycleChannel.send("AppLifecycleState.inactive");
```
Comment (xster) [inline shell/platform/android/io/flutter/view/FlutterView.java]: Might be worth testing, I might be wrong, but I think on most Android manufacturer OSes, the lock screen might put us in this state too. We should be sure about battery usage for common cases.

Comment (jonahwilliams) [reply shell/platform/android/io/flutter/view/FlutterView.java]: I'll take a look at battery use. I'm fairly certain that Choreographer won't give use frames, so hopefully it looks okay.

Comment (jonahwilliams) [reply shell/platform/android/io/flutter/view/FlutterView.java]: Did some research - we might be able to tell using the PowerManager API

Comment (jonahwilliams) [reply shell/platform/android/io/flutter/view/FlutterView.java]: I checked a few devices (pixel 1, nexus 5, galaxy s5) - all will call onStop when exiting to the lock screen.  I think this should be okay as is

**Comment by jonahwilliams** [comment]
Updates:

These changes work better with the engine/shell refactor patched in.  There are still some edge cases which happen if the app restarts when entering split screen.  We seem to get one or two extra frames after the view has been detached, might be a race in the teardown.  Also the video player plugin might have some unrelated issues on the android side.

All these issues are fixed if we add density and screenLayout to config changes. This would be a major breaking change, everyone would have to update their androidManifest, though they should anyway so their apps don't restart.

Thoughts @Hixie , @jason-simmons ?

**Comment by Hixie** [comment]
Looks like we already updated the manifest in https://github.com/flutter/flutter/pull/15871, in fact.

**Comment by jonahwilliams** [comment]
I've confirmed that this works on the latest flutter version locally, including video

**Pull request merged**